### ISSUE TITLE
How do we integrate custom gui functionality in the plugin?

### ISSUE THREAD

**Issue by constantinpape** [hci-unihd/covid-if-annotations#3]
Integrating the I/O function was straight-forward via the `@napari_hook_implementation` decorator.
(Great job!).
However, I am not quite sure what's the best way to integrate additional GUI functionality within the plugin, e.g. hiding a segment that was proofread, see #2 or going from step 1 (correcting the segmentation) to step 2 (classifying cells), see #1.

**Comment by sofroniewn** [comment]
Yes, this needs further discussion. I will open a dedicated issue on the napari repo and link here

**Comment by sofroniewn** [comment]
One approach would be for napari to provide additional hookspecs for keybindings, mouse functions, and GUI elements (either from Qt, or from the [magicgui](https://magicgui.readthedocs.io/en/latest/) library). This is the direction we want to go in, and what the issue I made proposes, but it will take time.

The other approach is that in the short term instead of being a true "napari plugin" (which would not need to depend on napari and just leverage hook_implementations) you make a package which minimally depends on napari and does a bit of the wiring up yourself.

You can see an example of that here [segmentify](https://github.com/transformify-plugins/segmentify) which was the summer project a very talented intern of mine worked on last year. The tool has similar functionality to ilastik, but was really just proof of concept for some interface ideas.

You'll see that segmentify depends on napari and that in its [init](https://github.com/transformify-plugins/segmentify/blob/master/segmentify/__init__.py#L4-L9) it imports and wires up keybindings defined in a dedicated [_key_bindings.py](https://github.com/transformify-plugins/segmentify/blob/master/segmentify/_key_bindings.py) file.

Segmentify also uses magicgui to create its own very simple widgets in a [gui.py file](https://github.com/transformify-plugins/segmentify/blob/master/segmentify/gui.py) - I highly recommend using magicgui for simple widgets like this. It is made by @tlambert03 from the napari team and if is very nice for the simple stuff (If you need more complex custom Qt widgets you can make your own).

To wire up segmentify you can see the [launch.py](https://github.com/transformify-plugins/segmentify/blob/master/examples/launch.py) example where the magicgui widget gets instantiated and then added through `viewer.window.add_dock_widget(gui)`

Basically in the short-term the approach like segmentify gives you lots of flexibility, but is more like creating your own "app" rather than strictly being a "plugin", but we havn't got all the hookspecs done yet for the full "plugin" interface to be able to recreate all the functionality that a tool like segmentify needs. Does that make sense?

**Comment by constantinpape** [comment]
> Basically in the short-term the approach like segmentify gives you lots of flexibility, but is more like creating your own "app" rather than strictly being a "plugin", but we havn't got all the hookspecs done yet for the full "plugin" interface to be able to recreate all the functionality that a tool like segmentify needs. Does that make sense?

Yes, that makes a lot of sense and I agree this is the way to go in the short-term.
I will look at the segmentify example later and see if I can set it up for our purpose.
Would the napari IO hooks would still work for this?

**Comment by constantinpape** [comment]
@sofroniewn I had a look at segmentify now and it was easy to adapt this to what we need here.
I think the main structure of the tool is there now.

Now I need to connect the pieces and hook up the points layer so it can be used for the semantic annotations. @kevinyamauchi already gave me some pointers on this in #2 and I think I should be able to implement it.
Also, I will try to implement the hiding functionality with magicgui.

There are still a couple of other minor points where I could use some advice:
- The reader hook is working as expected. However, I need to call some initialize function after it has been called, [e.g. for this](https://github.com/hci-unihd/covid-if-annotations/blob/master/napari_covid_if_annotations/launcher/covid_if_annotations.py#L32). Is there a way to have a listener for the reader hook that can trigger this?
- I have implemented a keybinding to update the points and edges from the segmentation, in order to reflect changes that have been done via painting, [see here](https://github.com/hci-unihd/covid-if-annotations/blob/master/napari_covid_if_annotations/_key_bindings.py#L11). Another possibility would be to have a listener for this, that triggers the update whenever a paint action is performed. Is this possible? (I don't actually know if it's a good idea, because the call will take ~1s so it might make painting too unresponsive). If we don't have a listener, it would be good to also have a gui button for this. Would that be possible via magicgui?

**Comment by sofroniewn** [comment]
> The reader hook is working as expected. However, I need to call some initialize function after it has been called, e.g. for this. Is there a way to have a listener for the reader hook that can trigger this?

Is `infected_labels` something that gets used across the `viewer` or just on one particular layer? If on one particular layer then each `add_*` method also takes an optional dictionary called `metadata` where you can put whatever you want, and that might be a good place to put this array. For example

```python
    viewer.add_labels(infected_edges, name='infected-classification', visible=False, metadata={'infected_labels': infected_labels}
```
 
We could add a global metadata dict to the viewer too
```python
napari.Viewer(metadata={'infected_labels': infected_labels})
```
or
```python
viewer.metadata = {'infected_labels': infected_labels}
```

I don't think we have a way to add a listener to a reader hook right now, maybe @tlameber03 knows.

> I have implemented a keybinding to update the points and edges from the segmentation, in order to reflect changes that have been done via painting, see here. Another possibility would be to have a listener for this, that triggers the update whenever a paint action is performed. Is this possible? (I don't actually know if it's a good idea, because the call will take ~1s so it might make painting too unresponsive).

There are some events that each layer emits - for example `events.set_data` which you could connect too. We don't have specific events for when painting gets finished say, but I've been wanting to add those - see https://github.com/napari/napari/issues/517 for an old issue mentioning that. This could be a good opportunity to figure out the best way to do that.

> If we don't have a listener, it would be good to also have a gui button for this. Would that be possible via magicgui?

Yes I think so, you should be able to get the needed layers passed to the function when the button is clicked and edit the properties of the layers you need to in place.

**Comment by constantinpape** [comment]
Ok, I was not aware of the layer metadata, that solves the initialization issues in a very clean way.
So also no need for a listener to the reader hook. Thanks!

> There are some events that each layer emits - for example `events.set_data` which you could connect too. We don't have specific events for when painting gets finished say, but I've been wanting to add those - see [napari/napari#517](https://github.com/napari/napari/issues/517) for an old issue mentioning that. This could be a good opportunity to figure out the best way to do that.

Ok, I am happy to give it a try if you happen to implement it, but if I can connect the keybinding to the GUI it's not really necessary here. I will try it via magicgui tomorrow.

**Comment by tlambert03** [comment]
> I don't think we have a way to add a listener to a reader hook right now, maybe @tlambert03 knows.

Sounds like you got this fixed in a much better way using `layer.metadata`... but just for the sake of completeness:  if you're looking to be notified whenever `napari_get_reader` is called, @sofroniewn is right, there's no "conventional" event listener on the reader hook.  However, it's worth pointing out that it's _your_ function, so _you_ know exactly when it's called! 😄 
you could put your own logic [right before returning layer data](https://github.com/hci-unihd/covid-if-annotations/blob/bfa541fe2595f5c02b6a42f1115c9ea37f34e0dc/napari_covid_if_annotations/io_hooks.py#L82) that handles/stores the `"infected_labels"` how you want them.

But again... `layer.metadata` is way better.

**Comment by constantinpape** [comment]
Thanks for explaining @tlambert03. But yeah, using `layer.metadata` is much cleaner in this case :).

**Comment by constantinpape** [comment]
@tlambert03 @sofroniewn 
I tried to use `magicgui` now to also create gui elements for the keybindings for updating the layers `u` / hiding annotated segments `h`. While this is redundant, I think it will be helpful for users to see this functionality as gui elements as well.

I have based this on the `segmentify` example. Here are the gui functions:
https://github.com/hci-unihd/covid-if-annotations/blob/master/napari_covid_if_annotations/gui.py
and here is the docking to the viewer (commented because it doesn't work yet, see below):
https://github.com/hci-unihd/covid-if-annotations/blob/master/napari_covid_if_annotations/launcher/covid_if_annotations.py#L41-L47 

This fails with the following error:
```
Traceback (most recent call last):
  File "launch.py", line 22, in <module>
    launch_covid_if_annotation_tool(path)
  File "/home/pape/Work/covid/covid-if-annotations/napari_covid_if_annotations/launcher/covid_if_annotations.py", line 47, in launch_covid_if_annotation_tool
    hide_gui = toggle_hide_annotated_segments_gui.Gui()
  File "/home/pape/Work/software/conda/miniconda3/envs/test-annotations/lib/python3.8/site-packages/magicgui/core.py", line 635, in __init__
    super().__init__(
  File "/home/pape/Work/software/conda/miniconda3/envs/test-annotations/lib/python3.8/site-packages/magicgui/core.py", line 187, in __init__
    self.set_widget(
  File "/home/pape/Work/software/conda/miniconda3/envs/test-annotations/lib/python3.8/site-packages/magicgui/core.py", line 336, in set_widget
    raise TypeError(msg)
TypeError: Unable to find the appropriate widget for function "toggle_hide_annotated_segments_gui", arg "viewer", type "<class 'napari.viewer.Viewer'>".
```
So it looks like passing `Viewer` to the function decorated with `magicgui` is currently not supported.

Let me know if there is a better way to achieve what I want to do or if this is something you consider fixing on the `magicgu` / `napari` side.

**Comment by tlambert03** [comment]
Hi @constantinpape, by default, magicgui wants to represent _every_ parameter in the function signature with a widget in the gui...  If you want it to skip any parameters, you can use the `ignore` option:

```python
# only arg1 will be in the gui
@magicgui(ignore=["arg2"])
def function(arg1, arg2):
    ...
```

However, in this case, what exactly are you hoping magicgui will show for the "viewer" parameter here?
```python
@magicgui(call_button='update layers [u]')
def update_layers_gui(viewer: Viewer):
    update_layers(viewer)
```
are there multiple viewers and you want the user to pick one? or do you literally just want a call button that acts as a shortcut for what already exists as a keybinding to a global viewer instance?

this might be a bit of a forced use case for magicgui... which is intended to represent _parameters_ in a function signature ([example](https://magicgui.readthedocs.io/en/latest/examples/napari_img_math/)).  If all you need is a single call button and you don't need to get any parameters from the user, then it's probably easier to just make a QPushButton directly?

**Comment by tlambert03** [comment]
another general point here about magicgui:  it's core feature is to map argument types to GUI widgets (i.e. an integer to a SpinBox, a string to a LineEdit, etc...).   So the `TypeError` you saw there is magicgui saying "I don't know what you want me to do with something of type `Viewer`".

While napari has provided some default behavior for objects annotated as `Layer` types (using the [register_type](https://magicgui.readthedocs.io/en/latest/type_inference/#register_type) hook from magicgui), there is no default behavior for arguments annotated as `Viewer` types. 

i'm also realizing... that `@Viewer.bind_key('u')` decorator there is doing some rather fancy object binding that makes it possible for it to pass your callback a specific viewer.

If we wanted to provide similar functionality to what the napari `bind_key` decorator is doing there within `magicgui` (namely, pass you the current viewer instance whenever you call the function), then we'd need to use `magicgui.register_type(napari.viewer.Viewer, ...)` and then figure out how to grab the current viewer instance (which, actually, that bind_key decorator is actually uniquely suited to do! 😂).  might need to think about this one...  for a bit

**Comment by constantinpape** [comment]
>  If you want it to skip any parameters, you can use the `ignore` option:

I added `ignore=['viewer']` to the decorator now. It starts napari now, but when I press one of the buttons, I get:
```
WARNING: Traceback (most recent call last):
  File "/home/pape/Work/software/conda/miniconda3/envs/test-annotations/lib/python3.8/site-packages/magicgui/core.py", line 199, in <lambda>
    self.call_button.clicked.connect(lambda checked: self.__call__())
  File "/home/pape/Work/software/conda/miniconda3/envs/test-annotations/lib/python3.8/site-packages/magicgui/core.py", line 523, in __call__
    value = self.func(**_kwargs)
TypeError: update_layers_gui() missing 1 required positional argument: 'viewer'
Aborted (core dumped)
```
Which I guess makes sense; it doesn't know that it should pass the viewer to the function.

> are there multiple viewers and you want the user to pick one? or do you literally just want a call button that acts as a shortcut for what already exists as a keybinding to a global viewer instance?

There is only one viewer. So yes, I just want a call button for the global viewer instance ;).

> this might be a bit of a forced use case for magicgui... which is intended to represent _parameters_ in a function signature ([example](https://magicgui.readthedocs.io/en/latest/examples/napari_img_math/)). If all you need is a single call button and you don't need to get any parameters from the user, then it's probably easier to just make a QPushButton directly?

I agree, this is a bit of a weird use case for `magicgui` (I also thought about this when I read the `magicgui` examples). But it would still be nice to be able to do this without the need to go to writing something in pyqt.

Also, for the `hide annotated segments` button, there is actually a bool flipped eventually when the shortcut or button is called, see https://github.com/hci-unihd/covid-if-annotations/blob/master/napari_covid_if_annotations/_key_bindings.py#L80.
I just wasn't sure what's the best way to expose this via the function signature, because this variable is stored in the metadata of some layer.

**Comment by constantinpape** [comment]
> If we wanted to provide similar functionality to what the napari `bind_key` decorator is doing there within `magicgui` (namely, pass you the current viewer instance whenever you call the function), then we'd need to use `magicgui.register_type(napari.viewer.Viewer, ...)` and then figure out how to grab the current viewer instance (which, actually, that bind_key decorator is actually uniquely suited to do! joy). might need to think about this one... for a bit

Ok, I think it would be quite cool to support it :). But for the project at hand it's not so important.
As you said, I could just implement a `QPushButton` instead.

**Comment by tlambert03** [comment]
> Ok, I think it would be quite cool to support it

yeah, I agree!  I'll keep thinking about it.  I think it will require a bit of a change to napari itself...

**Comment by tlambert03** [comment]
OK, here's one "solution" for this that you can implement without waiting for napari (though it has a downside)...
Add this to your code:

```python
from magicgui import register_type

def get_viewers(gui, *args):
    try:
        return (gui.parent().qt_viewer.viewer,)
    except AttributeError:
        return tuple(v for v in globals().values() if isinstance(v, napari.Viewer))

register_type(napari.Viewer, choices=get_viewers)
```

then you should be able to annotate an argument as a napari.Viewer, and it will pass you the viewer instance that the widget is currently embedded in):

```python
@magicgui(call_button='update layers [u]')
def update_layers_gui(viewer: Viewer):
    update_layers(viewer)
```

... the "downside" is that for the moment, it's going to show you a rather ugly `__repr__` of the viewer as dropdown:
<img width="657" alt="Screen Shot 2020-05-24 at 7 10 08 PM" src="https://user-images.githubusercontent.com/1609449/82767040-53b15b00-9df2-11ea-9b8b-f1c22b08f8fa.png">

We can make that repr prettier by adding a `__str__` method to napari.Viewer, but ideally it should be hideable.  That will require a change to magicgui though, because the "ignore" parameter is not quite as useful (for the "missing 1 required positional argument: 'viewer'" reason you saw above)

**Comment by tlambert03** [comment]
ok, I just pushed a new release of magicgui ([v0.1.5](https://pypi.org/project/magicgui/)) that lets you hide a widget.  I've also created a PR at napari (https://github.com/napari/napari/pull/1279) to register the `Viewer` type.  Until the napari PR merges, you can do the same thing locally by adding that `register_type` code in my previous example.  Then update to magicgui >= 0.1.5 and add the `visible` argument to your decorator:
```python
@magicgui(call_button='update layers [u]', viewer={"visible": False})
def update_layers_gui(viewer: Viewer):
    update_layers(viewer)
```

**Comment by constantinpape** [comment]
Works exactly as it should!
Thanks so much @tlambert03.

**Comment by constantinpape** [comment]
@tlambert03 one minor thing:
the initial layout of the gui elements is a bit weird; `update layers` is very large and `hide annotations` very small. Is there a way too easily change this, maybe with some argument to the decorator?

No worries if not, this is really just the icing on the cake. 

![Screenshot from 2020-05-25 10-36-20](https://user-images.githubusercontent.com/4263537/82795185-ee05b300-9e73-11ea-8235-be617100dd08.png)

**Comment by tlambert03** [comment]
This can be done, but it's really the main napari window you need to alter at that point (the magicgui widgets will just stretch to fill the size of their dockwidgets, and napari owns the dockwidgets).  The trick is to get a list of dockwidgets, then resize them, something like this (in your `connect_to_viewer`):
```python
    # hang on to dock_widget pointers
    dw0 = viewer.window.add_dock_widget(update_gui)
    dw1 = viewer.window.add_dock_widget(hide_gui)
    dw2 = viewer.window.add_dock_widget(paint_gui)

    # call resizeDocks... play with the numbers in the second list until you're happy
    from qtpy.QtCore import Qt
    viewer.window._qt_window.resizeDocks([dw0, dw1, dw2], [800, 800, 600], Qt.Horizontal)
```

**Comment by constantinpape** [comment]
Thanks!
I will start optimizing this once all the gui elements are added.

**Comment by sofroniewn** [comment]
@constantinpape as I noted in https://github.com/hci-unihd/covid-if-annotations/issues/9#issuecomment-633659801 you might want to combine all your custom widgets into one nice one and then add it to the viewer as a single dock widget - will give you more flexibility. You could have it vertically oriented and end up on the right hand side `viewer.window.add_dock_widget(custom_gui, area='right', allowed_areas=['right', 'left'])` That can end up looking quite nice instead of the long horizontal stuff at the bottom.

**Comment by constantinpape** [comment]
@sofroniewn 

> You could have it vertically oriented and end up on the right hand side `viewer.window.add_dock_widget(custom_gui, area='right', allowed_areas=['right', 'left'])` That can end up looking quite nice instead of the long horizontal stuff at the bottom.

This makes a lot of sense. I would need to combine [all the gui elements](https://github.com/hci-unihd/covid-if-annotations/blob/master/napari_covid_if_annotations/gui.py#L59-L76), i.e. the gui elements created by `magicgui` and `QLabel`, into `custom_gui` though and I don't know how to do this.

**Comment by sofroniewn** [comment]
You'd need to create a [QVBoxLayout](https://doc.qt.io/qt-5/qvboxlayout.html) to contain them and then set a [QWidget](https://doc.qt.io/qt-5/qwidget.html) to have that layout and then add that widget. I can provide more guidance if needed

**Comment by tlambert03** [comment]
just to add to this... in case it wasn't apparent, the `your_function.Gui` attribute is always a subclass of `QWidget`... and can be assembled in a layout the "regular qt" way as @sofroniewn described... something like:
```python
from qtpy.QtWidgets import QWidget, QVBoxLayout

w1 = your_function1.Gui()
w2 = your_function2.Gui()

widget = QWidget()
layout = QVBoxLayout()
widget.setLayout(layout)
layout.addWidget(w1)
layout.addWidget(w2)

viewer.window.add_dock_widget(widget)
```

**Comment by sofroniewn** [comment]
I wonder @tlambert03 would we ever add convenience functionality to `add_dock_widget` such that is someone passed a list of widgets, and maybe a flag for horizontal/ vertical we would do that composition under the hood for them? Could be very nice, or maybe it's a bit too magic? 

Alternatively magicgui could have a function called "compose_widgets" that did this, and maybe had more kwargs as needed?

**Comment by tlambert03** [comment]
yeah definitely.  both things.  (particularly the list of widgets to add_dock_widget)

**Comment by constantinpape** [comment]
It works and looks much better!
Thanks for your pointers @tlambert03 and @sofroniewn.

I still need to improve the tooltip, see #9, then I will share a screenshot of the current version.

**Comment by constantinpape** [comment]
This is how it looks right now. Much better than before, but still some room for improvements.

![Screenshot from 2020-05-27 13-58-39](https://user-images.githubusercontent.com/4263537/83016435-38c52d80-a022-11ea-8a5d-4bbd501fe12e.png)

**Comment by tlambert03** [comment]
you should be able to restrict the total height on the QWidget or the layout... which might improve things.  I'll try to take a look later today

**Comment by constantinpape** [comment]
@tlambert03 @sofroniewn After adding the magicgui buttons to the `QVBox` they don't work any more:
when I click them, the function gets passed `None` instead of the viewer instance.

See the code where I do this here:
https://github.com/hci-unihd/covid-if-annotations/blob/master/napari_covid_if_annotations/gui.py#L73-L84

**Comment by sofroniewn** [comment]
yes, sorry that should be fixed by https://github.com/napari/napari/pull/1296 which i am about to merge

**Comment by sofroniewn** [comment]
see https://github.com/napari/napari/pull/1296#issuecomment-634765757

**Comment by constantinpape** [comment]
@sofroniewn I have updated to the new napari release now, but I still see the same issue:
Clicking any of the buttons added via the `QVBox` crashes the viewer because `None` is passed instead of the viewer instance.

**Comment by tlambert03** [comment]
Hey @constantinpape.  you would actually need to _remove_ the whole `register_type(napari.Viewer, choices=get_viewers)` segment, (it's now part of napari and your section there is now overriding the update napari code)... however, it looks like there's still a bug even if you do it... sorry about that :(

I think that magicgui is not the right tool for this... way too many levels of indirection for what was already a simple gui task :). I just pushed #19 that shows how I'd do it without magicgui.

**Comment by constantinpape** [comment]
> I think that magicgui is not the right tool for this... way too many levels of indirection for what was already a simple gui task :). I just pushed #19 that shows how I'd do it without magicgui.

Thanks @tlambert03. I merged the PR and it works just fine.
I agree, the code for it is so simple, no need to support it via magicgui.

(Just in general having developer docs to show this would be great, but I know that is a ton of work...)

**Comment by tlambert03** [comment]
yeah ... an intro to extending the gui with qt widgets is sorely needed!  one of these days :)

**Comment by constantinpape** [comment]
Thanks for all the feedback @tlambert03 @sofroniewn. The GUI is in a state now where we are happy with it for the time being.
